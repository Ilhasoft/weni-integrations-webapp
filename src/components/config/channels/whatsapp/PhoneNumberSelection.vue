<template>
  <unnnic-modal
    ref="phone-number-selection-modal"
    class="phone-number-selection"
    :text="$t('WhatsAppCloud.config.phone_numbers.connect')"
    scheme="feedback-green"
    modal-icon="phone-3"
    @close="closePopUp"
    @click.stop
  >
    <skeleton-loading v-if="loadingPhoneNumbers || loadingDebugToken" slot="message" />
    <div v-else slot="message">
      <unnnic-select
        :search="false"
        size="sm"
        v-model="selectedNumber"
        placeholder="Select your phone number"
      >
        <option v-for="(number, index) in whatsAppPhoneNumbers" :key="index">
          {{ number.verified_name }} ~ ({{ number.display_phone_number }})
        </option>
      </unnnic-select>

      <div class="phone-number-selection__buttons">
        <unnnic-button
          class="phone-number-selection__buttons__cancel"
          type="terciary"
          size="large"
          :text="$t('WhatsAppCloud.config.phone_numbers.connect_later')"
          @click="closePopUp"
        ></unnnic-button>

        <unnnic-button
          class="phone-number-selection__buttons__save"
          type="terciary"
          size="large"
          :text="
            loadingWhatsAppCloudConfigure
              ? $t('general.loading')
              : $t('WhatsAppCloud.config.phone_numbers.create_channel')
          "
          @click="createChannel"
          :iconLeft="loadingWhatsAppCloudConfigure ? 'loading-circle-1' : null"
        ></unnnic-button>
      </div>
    </div>
  </unnnic-modal>
</template>

<script>
  import { mapActions, mapState, mapGetters } from 'vuex';
  import skeletonLoading from './loadings/PhoneNumberSelection.vue';
  import { unnnicCallAlert } from '@weni/unnnic-system';

  export default {
    name: 'PhoneNumberSelection',
    components: {
      skeletonLoading,
    },
    props: {
      app: {
        type: Object,
        default: /* istanbul ignore next */ () => {},
      },
      customData: {
        type: Object,
        default: /* istanbul ignore next */ () => {},
      },
    },
    data() {
      return {
        showModal: false,
        selectedNumber: null,
      };
    },
    beforeDestroy() {
      this.setSelectedPhoneNumber({ data: null });
    },
    /* istanbul ignore next */
    async mounted() {
      await this.fetchDebugToken();
      if (!this.errorDebugToken) {
        await this.fetchPhoneNumbers();
      }
    },
    computed: {
      ...mapGetters(['getSelectedProject']),
      ...mapState('WhatsAppCloud', [
        'loadingPhoneNumbers',
        'loadingDebugToken',
        'loadingWhatsAppCloudConfigure',
        'errorCloudConfigure',
        'errorPhoneNumbers',
        'errorDebugToken',
        'selectedPhoneNumber',
        'wabaId',
        'businessId',
        'whatsAppPhoneNumbers',
      ]),
    },
    methods: {
      ...mapActions({
        setSelectedPhoneNumber: 'WhatsAppCloud/setSelectedPhoneNumber',
        configurePhoneNumber: `WhatsAppCloud/configurePhoneNumber`,
        getDebugToken: 'WhatsAppCloud/getDebugToken',
        getWhatsAppPhoneNumbers: 'WhatsAppCloud/getWhatsAppPhoneNumbers',
      }),
      closePopUp() {
        this.showModal = !this.showModal;
        this.$emit('closePopUp');
      },
      async createChannel() {
        if (!this.loadingWhatsAppCloudConfigure) {
          const data = {
            waba_id: this.wabaId,
            business_id: this.businessId,
            phone_number_id: this.selectedPhoneNumber?.id,
            input_token: this.customData.input_token,
            project_uuid: this.getSelectedProject,
          };

          await this.configurePhoneNumber({ data });

          if (this.errorCloudConfigure) {
            this.callErrorModal({
              text: this.$t('WhatsAppCloud.config.configure_phone_number.error'),
            });
          } else {
            this.$router.replace('/apps/my');
          }
        }
      },
      async fetchDebugToken() {
        const params = {
          input_token: this.customData.input_token,
        };

        await this.getDebugToken({ params });

        if (this.errorDebugToken) {
          this.callErrorModal({
            text: this.$t('WhatsAppCloud.config.debug_token.error'),
          });
        }
      },
      async fetchPhoneNumbers() {
        const params = {
          waba_id: this.wabaId,
          input_token: this.customData.input_token,
        };

        await this.getWhatsAppPhoneNumbers({ params });

        if (this.errorPhoneNumbers) {
          this.callErrorModal({
            text: this.$t('WhatsAppCloud.config.phone_numbers.error'),
          });
        }
      },
      callErrorModal({ text }) {
        unnnicCallAlert({
          props: {
            text: text,
            title: 'Error',
            icon: 'alert-circle-1',
            scheme: 'feedback-red',
            position: 'bottom-right',
            closeText: this.$t('general.Close'),
          },
          seconds: 6,
        });
      },
    },
    watch: {
      selectedNumber(newValue) {
        const newValueDisplayNumber = newValue
          .split('~')[1]
          .replaceAll('(', '')
          .replaceAll(')', '')
          .trim();

        const number = this.whatsAppPhoneNumbers.find(
          (number) => number.display_phone_number === newValueDisplayNumber,
        );

        this.setSelectedPhoneNumber({ data: number });
      },
    },
  };
</script>

<style lang="scss" scoped>
  .phone-number-selection {
    cursor: default;

    &__buttons {
      display: flex;
      justify-content: space-around;
      margin-top: $unnnic-spacing-stack-lg;

      &__save {
        ::v-deep svg {
          animation: rotation 1.5s infinite linear;
        }
      }
    }

    ::v-deep .unnnic-modal-container-background-body-description {
      padding-bottom: $unnnic-spacing-stack-lg;
    }
  }

  @keyframes rotation {
    from {
      transform: rotate(0deg);
    }
    to {
      transform: rotate(360deg);
    }
  }
</style>
