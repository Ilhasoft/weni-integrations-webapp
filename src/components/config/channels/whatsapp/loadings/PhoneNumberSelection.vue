<template>
  <div class="whatsapp-phone-number-loading">
    <unnnic-skeleton-loading
      class="whatsapp-phone-number-loading__content"
      tag="div"
      height="38px"
    />
    <div class="whatsapp-phone-number-loading__buttons">
      <unnnic-button
        class="whatsapp-phone-number-loading__buttons__cancel"
        type="terciary"
        size="large"
        :text="$t('WhatsAppCloud.config.phone_numbers.connect_later')"
      ></unnnic-button>

      <unnnic-button
        class="whatsapp-phone-number-loading__buttons__save"
        type="terciary"
        size="large"
        :text="$t('general.loading')"
        iconLeft="loading-circle-1"
      ></unnnic-button>
    </div>
  </div>
</template>

<script>
  export default {};
</script>

<style lang="scss" scoped>
  @keyframes rotation {
    from {
      transform: rotate(0deg);
    }
    to {
      transform: rotate(360deg);
    }
  }

  .whatsapp-phone-number-loading {
    &__content {
      height: $unnnic-avatar-size-sm;
    }

    &__buttons {
      display: flex;
      justify-content: space-around;
      margin-top: $unnnic-spacing-stack-lg;

      &__save {
        ::v-deep svg {
          animation: rotation 1.5s infinite linear;
        }
      }
    }
  }
</style>
