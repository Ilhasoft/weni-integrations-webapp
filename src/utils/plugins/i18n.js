import Vue from 'vue';
import VueI18n from 'vue-i18n';

Vue.use(VueI18n);

import en from '../../locales/en.json';
import pt_br from '../../locales/pt_br.json';
import es_es from '../../locales/es_es.json';

const languages = {
  'en-us': en,
  'pt-br': pt_br,
  es: es_es,
};

const messages = Object.assign(languages);

const i18n = new VueI18n({
  locale: 'en-us',
  fallbackLocale: 'en-us',
  messages,
});

export default i18n;
