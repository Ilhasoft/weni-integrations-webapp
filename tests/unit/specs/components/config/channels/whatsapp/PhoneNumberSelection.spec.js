import VueRouter from 'vue-router';
import Vuex from 'vuex';
import { shallowMount, createLocalVue } from '@vue/test-utils';
import PhoneNumberSelection from '@/components/config/channels/whatsapp/PhoneNumberSelection.vue';
import i18n from '@/utils/plugins/i18n';
import { singleApp } from '../../../../../../__mocks__/appMock';

const router = new VueRouter();

const localVue = createLocalVue();
localVue.use(VueRouter);
localVue.use(Vuex);

describe('whatsapp/PhoneNumberSelection.vue', () => {
  let wrapper;
  let actions;
  let wppCloudGetters;
  let wppCloudActions;
  let store;

  beforeEach(() => {
    const customData = {
      pages: [
        { id: 1, name: 'page1' },
        { id: 2, name: 'page2' },
      ],
    };

    actions = {
      getApp: jest.fn(() => {
        return { data: singleApp };
      }),
      fetchWppProfile: jest.fn(() => {
        return { data: { photo_url: 'photo' } };
      }),
    };

    wppCloudGetters = {
      whatsAppPhoneNumbers: jest.fn(() => {
        return [];
      }),
      selectedPhoneNumber: jest.fn(() => {
        return { data: {} };
      }),
      wabaId: jest.fn(() => {
        return { data: {} };
      }),
      businessId: jest.fn(() => {
        return { data: {} };
      }),
    };

    wppCloudActions = {
      setSelectedPhoneNumber: jest.fn(),
      configurePhoneNumber: jest.fn(),
    };

    store = new Vuex.Store({
      modules: {
        WhatsAppCloud: {
          namespaced: true,
          actions: wppCloudActions,
          getters: wppCloudGetters,
        },
      },
      actions,
    });

    wrapper = shallowMount(PhoneNumberSelection, {
      localVue,
      i18n,
      router,
      store,
      propsData: {
        app: singleApp,
        customData,
      },
      mocks: {
        $t: () => 'some specific text',
        $route: {
          path: '/apps/1/details',
        },
      },
      stubs: {
        UnnnicModal: true,
        UnnnicButton: true,
        UnnnicSelect: true,
      },
    });
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should be rendered properly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('closePopUp()', () => {
    it('should set showModal opposite', () => {
      wrapper.setData({ showModal: true });
      expect(wrapper.vm.showModal).toBeTruthy();
      wrapper.vm.closePopUp();
      expect(wrapper.vm.showModal).toBeFalsy();
    });

    it('should call parent closePopUp()', () => {
      expect(wrapper.emitted('closePopUp')).toBeFalsy();
      wrapper.vm.closePopUp();
      expect(wrapper.emitted('closePopUp')).toBeTruthy();
    });
  });

  describe('selectedNumberWatcher', () => {
    it('should call setSelectedPhoneNumber', async () => {
      expect(wppCloudActions.setSelectedPhoneNumber).not.toHaveBeenCalled();
      await wrapper.setData({ selectedNumber: 'number ~ (123)' });
      expect(wppCloudActions.setSelectedPhoneNumber).toHaveBeenCalledTimes(1);
    });
  });

  describe('beforeDestroy()', () => {
    it('should call setSelectedPhoneNumber', async () => {
      expect(wppCloudActions.setSelectedPhoneNumber).not.toHaveBeenCalled();
      await wrapper.destroy();
      expect(wppCloudActions.setSelectedPhoneNumber).toHaveBeenCalledTimes(1);
    });
  });

  describe('createChannel()', () => {
    it('should change url into /apps/my', async () => {
      expect(wrapper.vm.$route.path).not.toEqual('/apps/my');
      await wrapper.vm.createChannel();
      expect(wrapper.vm.$route.path).toEqual('/apps/my');
    });

    it('should call configurePhoneNumber', async () => {
      expect(wppCloudActions.configurePhoneNumber).not.toHaveBeenCalled();
      await wrapper.vm.createChannel();
      expect(wppCloudActions.configurePhoneNumber).toHaveBeenCalledTimes(1);
    });
  });
});
