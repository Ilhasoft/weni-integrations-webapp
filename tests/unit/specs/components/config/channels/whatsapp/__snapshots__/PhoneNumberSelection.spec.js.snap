// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`whatsapp/PhoneNumberSelection.vue should be rendered properly 1`] = `
<unnnicmodal-stub text="some specific text" scheme="feedback-green" modal-icon="phone-3" class="phone-number-selection">
  <div>
    <unnnicselect-stub size="sm" placeholder="Select your phone number"></unnnicselect-stub>
    <div class="phone-number-selection__buttons">
      <unnnicbutton-stub type="terciary" size="large" text="some specific text" class="phone-number-selection__buttons__cancel"></unnnicbutton-stub>
      <unnnicbutton-stub type="terciary" size="large" text="some specific text" class="phone-number-selection__buttons__save"></unnnicbutton-stub>
    </div>
  </div>
</unnnicmodal-stub>
`;
